File: pets/algorithms/graph.go
//Алгоритм поиска в ширину по графам
package main

import (
	"container/list"
	"fmt"
)

func isManager(name string) bool {
	//Тупо для теста, если имя заканчивается на м - то менеджер
	if name[len(name)-1:] == "m" {
		return true
	} else {
		return false
	}
}

func findManagerInGraph(graph map[string][]string) string {

	// инициализируем очередь
	queue := list.New()

	//Добавляем в очередь первичный графф, связи 1 уровня
	for _, value := range graph["you"] {
		queue.PushBack(value)
	}

	//Работаем с очередью
	for queue.Len() > 0 {
		e := queue.Front() // Первый элемент
		name := e.Value.(string)

		if isManager(name) {
			return name
			break
		} else {
			if len(graph[name]) > 0 {
				for _, value := range graph[name] {
					queue.PushBack(value)
				}
			}
		}
		queue.Remove(e)
	}

	return "Нету менеджеров"

}

func main() {

	graph := map[string][]string{}
	graph["you"] = []string{"alice", "bob", "claire"}
	graph["bob"] = []string{"anuj", "peggy"}
	graph["alice"] = []string{"peggy"}
	graph["claire"] = []string{"thom", "jonny"}
	graph["anuj"] = []string{}
	graph["peggy"] = []string{}
	graph["thom"] = []string{}
	graph["jonny"] = []string{}

	fmt.Println(findManagerInGraph(graph))

}

File: pets/algorithms/graph.php
<?php
//Реализуем очередь FIFO
class Queue {
  // Массив очереди
  public $data = [];

  // Получаем длину очереди
  public function getLength():int {
    return count($this->data);
  }
  // Определяем, пуста ли очередь
  public function isEmpty():bool {
    return $this->getLength() === 0;
  }

  //Добавляем в конец очереди
  public function pushBack($element) {
    $this->data[] = $element;
  }

  //Получаем первый элемент очереди и удаляем его из очереди
  public function getFront():string {
    return (string)array_shift($this->data);
  }

  //Вернуть все
  public function getAll():array {
    return $this->data;
  }

}


function isManager(string $name):bool {
  //Тупо для теста, если имя заканчивается на м - то менеджер
  if( substr($name, -1) == "m" ){
    return true;
  } else {
    return false;
  }
}


function findManagerInGraph(array $graph):string {
  // инициализируем очередь
  $q = new Queue();

  //Добавляем в очередь первичный графф, связи 1 уровня
  foreach ($graph["you"] as $key => $value) {
    $q->pushBack($value);
  }

  //Работаем с очередью
  while ( !$q->isEmpty() ) {
    $name = $q->getFront();

    if( isManager($name) ){
      return $name;
    } else {
      if( count($graph[$name]) > 0 ){
        foreach ($graph[$name] as $key => $value) {
          $q->pushBack($value);
        }
      }
    }

  }

  return "Нету менеджеров";
}


$graph = [];
$graph["you"] = ["alice","bob","claire"];
$graph["bob"] = ["anuj","peggy"];
$graph["alice"] = ["peggy"];
$graph["claire"] = ["thom", "jonny"];
$graph["anuj"] = [];
$graph["peggy"] = [];
$graph["thom"] = [];
$graph["jonny"] = [];

var_dump(findManagerInGraph($graph));
